### PR TITLE
Support new clang name mangling of main to __main_argv_argv

### PR DESCRIPTION
**Pull request by sbc100**
This change just ensure that the linker preserves main whichever
name it is known by.

See the corresponding binaryen and llvm changes:

https://github.com/WebAssembly/binaryen/pull/2671
https://reviews.llvm.org/D75277

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1385,6 +1385,10 @@ def is_supported_link_flag(f):
         # include it by default.  Should be eliminated by meta-DCE if unused.
         shared.Settings.EXPORTED_FUNCTIONS += ['_setThrew']
 
+    if shared.Settings.WASM_BACKEND:
+      if '_main' in shared.Settings.EXPORTED_FUNCTIONS:
+        shared.Settings.EXPORTED_FUNCTIONS += ['___main_argc_argv']
```
Comment (kripken) [inline emcc.py]: doesn't this need to depend on whether we are in standalone mode?

this adds an export - do we need to export both `_main` and `__main_argc_argv`? (if one calls the other, we just need the outer one I think)

Comment (sbc100) [reply emcc.py]: All this does right now export it *if* it exists.  Neither symbols need to actually exits, it just stops them from being DCE'd by the linker if they do exist.

Comment (kripken) [reply emcc.py]: Oh right, we only error on exports the user provides, not ones we add internally, thanks.

File: system/lib/libc/crt1.c [hunk 2]
```diff
@@ -9,67 +9,30 @@
...
     __wasm_call_ctors();
   }
 
-  if (!main) {
+  if (!main && !__main_argc_argv) {
     return;
   }
 
+  /*
+   * Will either end up call the user's original zero argument main directly
```
Comment (kripken) [inline system/lib/libc/crt1.c]: call => calling

File: system/lib/libc/crt1.c [hunk 3]
```diff
@@ -9,67 +9,30 @@
...
-#include <stdio.h>
 
-extern void __wasm_call_ctors(void) __attribute__((weak));
+__attribute__((weak)) void __wasm_call_ctors(void);
 
-// TODO(sbc): We shouldn't even link this file if there is no main:
-// https://github.com/emscripten-core/emscripten/issues/9640
-extern int main(int argc, char** argv) __attribute__((weak));
+__attribute__((weak)) int __main_argc_argv(int argc, char** argv);
+__attribute__((weak)) int main(int argc, char** argv);
```
Comment (kripken) [inline system/lib/libc/crt1.c]: maybe a comment here explaining what the two are?

File: system/lib/standalone/__main_argc_argv.c [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// This file should no longer be need once we land the llvm-side change to
+// swtich to using __main_argc_argv:
+// TODO(https://reviews.llvm.org/D75277)
+
+// See https://github.com/CraneStation/wasi-libc/pull/152
+
+#include <assert.h>
+
+// Main has to be weak here because in emscripten we allow standalone
+// application that don't contain a main at all.
```
Comment (kripken) [inline system/lib/standalone/__main_argc_argv.c]: application => applications

File: system/lib/standalone/__main_argc_argv.c [hunk 5]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2020 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+// This file should no longer be need once we land the llvm-side change to
```
Comment (kripken) [inline system/lib/standalone/__main_argc_argv.c]: need => needed

File: system/lib/standalone/__main_argc_argv.c [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// swtich to using __main_argc_argv:
```
Comment (kripken) [inline system/lib/standalone/__main_argc_argv.c]: swtich => switch

File: system/lib/standalone/__main_void.c [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ * found in the LICENSE file.
+ */
+
+#include <alloca.h>
+#include <wasi/api.h>
+#include <sysexits.h>
+
+int __main_argc_argv(int argc, char *argv[]);
+
+// If main() uses argc/argv, then no __original_main is emitted, and then
```
Comment (kripken) [inline system/lib/standalone/__main_void.c]: I thought `__original_main` is no longer emitted by llvm? Or does emitted here mean something else?

File: tests/other/metadce/mem_no_argv_O3_STANDALONE_WASM_flto.funcs [hunk 8]
```diff
@@ -1,3 +1,4 @@
+$__original_main
```
Comment (kripken) [inline tests/other/metadce/mem_no_argv_O3_STANDALONE_WASM_flto.funcs]: why was this added?

File: tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.funcs [hunk 9]
```diff
@@ -1,4 +1,5 @@
-$__wasm_call_ctors
+$__main_void
+$_start
```
Comment (kripken) [inline tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.funcs]: seems like before they were inlined into one function? or main was a duplicate of `wasm_call_ctors` (empty). So might be worth finding out why that changed.

File: tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports [hunk 10]
```diff
@@ -1 +1,3 @@
-
+args_get
+args_sizes_get
```
Comment (kripken) [inline tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports]: this seems wrong, if there is no main, we shouldn't use args_get?

File: tests/test_core.py [hunk 11]
```diff
@@ -351,6 +351,10 @@ def test_hello_world(self):
     src = open('src.c.o.js').read()
     assert 'EMSCRIPTEN_GENERATED_FUNCTIONS' not in src, 'must not emit this unneeded internal thing'
 
+  @also_with_standalone_wasm
+  def test_hello_argc(self):
+    self.do_run_in_out_file_test('tests', 'core', 'test_hello_argc')
```
Comment (kripken) [inline tests/test_core.py]: is a new test needed? a bunch of other tests happen to use argc/argv.

Comment (sbc100) [reply tests/test_core.py]: Yeah, I wanted a small targeted one that runs in standalone mode.  Maybe unnecessary, but this helped me while writing this patch.

File: tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports (hunk 10 above)
Comment (sbc100) [reply tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports]: So far I've not found a way to handle the following scenario:

1. You don't have a main
2. You are using STANDALONE_WASM mode
3. You want to avoid all the costs associationed with _start and argv etc.
4. You won't want to pass any extra linker flags to signal that you don't have a main

I'm kind of OK with dropping the transparent handling of this use case.  Because (a) WASI doesn't yet support programs without main (b) it seems reasonable that if you really care about this configuration we can expect you to pass something like `-no-entry`, `-nostdlib`, `-s EXPORT_FUNCTIONS` without main included, to the linker.

So I propose the we main linking without a main a linker error in WASM_STANDALONE mode and ask people to explicitly opt into this using EXPORTED_FUNCTIONS with main not present.. does that seems reasonable?

**Comment by sbc100** [comment]
This change is still a WIP..  still need to tweak some stuff to get it right.

File: tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports (hunk 10 above)
Comment (kripken) [reply tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports]: What concerns me is that the standalone case without main is basically the standard "plugin" case. So I actually think it's common, more common than running wasi in the shell.

>  ask people to explicitly opt into this using EXPORTED_FUNCTIONS with main not present.. does that seems reasonable?

That seems reasonable - it looks like main was the one exception to this rule actually, as we already error on anything else in the list that isn't actually present.

Comment (sbc100) [reply tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports]: Ah yes, I think I'm proposing that (for WASM_STANDALONE at least) we remove that exception and show an error such as `main symbol not defined, if that is intentional please build with flag XXX to build a program without a main`

Comment (kripken) [reply tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports]: Sounds good. Can the flag just be not having `_main` in EXPORTED_FUNCTIONS?

Comment (sbc100) [reply tests/other/metadce/mem_no_main_O3_STANDALONE_WASM.imports]: Yes

**Comment by sbc100** [comment]
The binaryen side has now landed.

I'd like to land this PR which paves the way for the LLVM side to land.
